packages/pyright-internal/src/languageService/hoverProvider.ts

```diff
@@ -482,10 +484,29 @@ export class HoverProvider {
     }
 
     private static _getTypeText(node: NameNode, evaluator: TypeEvaluator, expandTypeAlias = false): string {
-        const type = evaluator.getType(node) || UnknownType.create();
+        let type = evaluator.getType(node) || UnknownType.create();
+        type = this._limitOverloadBasedOnCall(node, evaluator, type);
         return ': ' + evaluator.printType(type, { expandTypeAlias });
     }
 
+    private static _limitOverloadBasedOnCall(node: NameNode, evaluator: TypeEvaluator, type: Type) {
```

Comment:
one refactoring idea is adding getBestMatchType or something in TypeUtils, that basically does

```
function getBestMatchType(evaluator: TypeEvaluator, node: NameNode): string {
        let type = evaluator.getType(node) || UnknownType.create();
        if (if (isOverloadedFunction(type)) {
            const callNode = ParseTreeUtils.getCallForName(node);
            if (callNode) {
                const callTypeResult = evaluator.getTypeResult(callNode);

                if (callTypeResult?.overloadsUsedForCall && callTypeResult.overloadsUsedForCall.length > 0) {
                    type = OverloadedFunctionType.create(callTypeResult.overloadsUsedForCall);
                }
            }
        }
        
        return type;
    }
```
and replace evaluator.getType(node) to TypeUtils.getBestMatchType(evaluator, node) for all places we want this behavior.

Reply:
TypeUtils is a utility module that is consumed by the TypeEvaluator. TypeUtils therefore cannot consume the TypeEvaluator without creating an inversion. If we were to factor this out, it would need to go into the TypeEvaluator itself. I don't think there's a good enough reason to do that right now.

Reply:
I didn't mean literally the TypeUtils file we have. it could be any other (or new file) util file. since I believe all others will basically need to do the same work if they do.

Reply:
It would help with inlay hints on call function parameters if I could match the overloads